sips/verifiable_consensus_data.md

```diff
@@ -0,0 +1,235 @@
+    DataSSZ []byte `ssz-max:"128"` // attestation data max size
+}
+```
+
+All Attestation and SyncCommittee logic should be moved to work with the new type.
+
+```go
+func (cd VerifiableConsensusData) GetAttestationData (att phase0.AttestationData,
+                                                     headProof phase0.SignedBeaconBlockHeader, 
+                                                     targetProof phase0.SignedBeaconBlockHeader, 
+                                                     sourceProof phase0.SignedBeaconBlockHeader, 
+                                                     error) {
+    if err := att.UnmarshalSSZ(cd.DataSSZ); err != nil {
+        return nil, nil, nil, nil, errors.Wrap(err, "could not unmarshal attestation ssz")
+    }
+    // Head vote proof
+    if err := headProof.UnmarshalSSZ(cd.Proof[0]); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal head vote proof ssz")
+    }
+    // Target proof
+    if err := targetProof.UnmarshalSSZ(cd.Proof[1]); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal target proof ssz")
+    }
+    // Target proof
+    if err := sourceProof.UnmarshalSSZ(cd.Proof[1]); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal source proof ssz")
+    }
+    return att, headProof, targetProof, sourceProof, nil
+}
+
+func (cd VerifiableConsensusData) GetSyncCommitteeBlockRoot (phase0.Root, phase0.SignedBeaconBlockHeader, error) {
+	ret := SSZ32Bytes{}
+	if err := ret.UnmarshalSSZ(cd.DataSSZ); err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal ssz")
+	}
+    proof := phase0.SignedBeaconBlockHeader
+    if err := proof.UnmarshalSSZ(cd.Proof); err != nil {
+        return nil, nil, errors.Wrap(err, "could not unmarshal proof ssz")
+    }
+	return phase0.Root(ret), proof, nil
+}
+
+func (cd VerifiableConsensusData) Validate error {
+    switch cid.Duty.Type {
+	case BNRoleAttester:
+            if _, _, _, _, err := cid.GetAttestationData(); err != nil {
+                return err
+            }
+            return nil
+	case BNRoleSyncCommittee:
+            if _, _, err := cid.GetSyncCommitteeBlockRoot(); err != nil {
+                return err
+            }
+            return nil
+        default:
+		return errors.New("unknown duty role")
+	}
+}
+```
+
+In addition we do the following changes to value checks mechanisms:
+```go 
+func verifyData(expectedRoot phase0.Root, proof phase0.SignedBeaconBlockHeader) error {
+    beaconRoot := proof.Message.HashRoot()
+    if !bytes.Equal(expectedRoot, beaconRoot) {
+        return errors.New("unexpected header root")
+    }
+    blockEpoch := getBeaconNode().getEpochBySlot(proof.Message.slot)
+    domain := GetBeaconNode().DomainData(blockEpoch, spec.DomainProposer)
+    root := spec.ComputeEthSigningRoot(beaconRoot, domain)
+    validator := GetBeaconNode.GetValidator("HEAD", proof.Message.ProposerIndex)
+    return bls.Verify(sig, validator.PubKey, root)
```

Comment:
With this check, we're sure that the block proposed as head exists. But we, still, can't be sure if a malicious operator is sending us an old block to attest, correct?
Perhaps, if we had some state that persisted through all duties, we could do a lower limit check for the slot/epoch. This is some future discussion but what do you think?

Reply:
We can compare against the block we got from our node simply without persisting state.
But I see it as a future improvement outside this SIP